rust/projects/parallelism/project.md

```diff
@@ -0,0 +1,447 @@
+`Clone` for `KvsStore`.
+
+At this point, if you've carried over your code from previous projects, your
+single-threaded `KvsServer` should work once again, but now with a `KvsEngine`
+that can later be shared across thrteads.
+
+**Test cases to complete**:
+
+  - TODO @Yilin
+
+
+## Part 3: Adding multithreading to `KvsServer`
+
+Let's quickly review our architecture here: `KvsServer` sets up a TCP socket and
+begins listening on it; when it receives a request it deserializes it and calls
+some implementation of the `KvsEngine` trait to store or retrieve data; then it
+sends back the response. The details of how `KvsEngine` works don't matter to
+`KvsServer`.
+
+So in the last project you probably created a loop vaguelly like:
+
+```rust
+let listener = TcpListener::bind("127.0.0.1:80")?;
+
+for stream in listener.incoming() {
+	let cmd = self.read_cmd(&stream);
+	let resp = self.process_cmd(cmd);
+	self.respond(&stream, resp);
+}
+```
+
+_Well, now you just need to do the same thing, but spawn all the work inside the
+loop into your `NaiveThreadPool`_. The database query and the response are both
+handled on a different thread than the TCP listener.
+
+**Test cases to complete**:
+
+  - TODO @Yilin
+
+Again, you should still have a working client/server key-value store, now
+multithreaded.
+
+
+## Part 4: Creating a real thread pool
+
+So now that you've got your multi-threaded architecture in place, it's time to
+write a real thread pool. You probably wouldn't write your own thread pool in
+practice as there exist thread pool crates that are well-tested, but it
+is a useful exercise to gain experience with parallelism in general. Later in
+this project you will, as we did with the engine in the previous project,
+abstract the thread pool and compare the performance of yours with an existing.
+
+So, what is a thread pool?
+
+It's nothing complicated. Instead of creating a new thread for every
+multi-threaded job to be performed, a thread pool maintains a "pool" of threads,
+and reuses those threads instead of creating a new one.
+
+But why?
+
+It's entirely about performance. It saves a small amount of performance, but
+when writing high-performance applications, every bit counts. Imagine what it
+takes to make a new thread:
+
+You've got to have a call stack for that thread to run on. That's an allocation.
+Allocations are pretty cheap, but not as cheap as not allocation. How that call
+stack is allocated depends on details of the operating system and runtime, but
+can involve locks and syscalls. Syscalls again are not _that_ expensive, but
+they are expensive when we're dealing with Rust levels of performance. Reducing
+syscalls is a common source of easy optimization. That stack then has to be
+carefully initialized so that first [stack _frame_] contains the appropriate
+values for the return pointer and other things. In Rust the stack needs to be
+configured with a [guard page]. That's two more syscalls, [to `mmap` and to
```

Comment:
Seems that these two syscalls are saved in Linux. Add "except in Linux" for strictness?

Reply:
:eyeroll: yeah, I saw that after I wrote this and read the std src. It's a bummer because I think this is an illustrative point.

It's also unfortunate that all the dirty mechanics of launching the thread are hidden in libpthread.

I'll do as you suggest.